OpenRA.Mods.Cnc/Traits/DrainPrerequisitePowerOnDamage.cs

```diff
@@ -0,0 +1,68 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[Desc("Converts damage to a charge level of a GrantPrerequisiteChargeDrainPower.")]
+	public class DrainPrerequisitePowerOnDamageInfo : ConditionalTraitInfo
+	{
+		[Desc("The OrderName of the GrantPrerequisiteChargeDrainPower to drain.")]
+		public readonly string OrderName = "GrantPrerequisiteChargeDrainPowerInfoOrder";
+
+		[Desc("Damage is multiplied by this number when converting damage to drain ticks.")]
+		public readonly int DamageMultiplier = 1;
+
+		[Desc("Damage is divided by this number when converting damage to drain ticks.")]
+		public readonly int DamageDivisor = 600;
+
+		public override object Create(ActorInitializer init) { return new DrainPrerequisitePowerOnDamage(init.Self, this); }
+	}
+
+	public class DrainPrerequisitePowerOnDamage : ConditionalTrait<DrainPrerequisitePowerOnDamageInfo>, INotifyOwnerChanged, IDamageModifier
+	{
+		SupportPowerManager spm;
+
+		public DrainPrerequisitePowerOnDamage(Actor self, DrainPrerequisitePowerOnDamageInfo info)
+			: base(info) { }
+
+		protected override void Created(Actor self)
+		{
+			base.Created(self);
+			spm = self.Owner.PlayerActor.Trait<SupportPowerManager>();
+		}
+
+		void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			spm = newOwner.PlayerActor.Trait<SupportPowerManager>();
+		}
+
+		int IDamageModifier.GetDamageModifier(Actor self, Damage damage)
+		{
+			if (!IsTraitDisabled && damage != null)
+			{
+				var damageSubTicks = (int)(damage.Value * 100L * Info.DamageMultiplier / Info.DamageDivisor);
```

Comment:
Guard for overflow?

Reply:
The overflow is already guarded by casting to a long with the `100L`.